defaults_test.go

```diff
@@ -0,0 +1,307 @@
+// Author: Bodie Solomon
+//         bodie@synapsegarden.net
+//         github.com/binary132
+//
+//         2015-02-16
+
+package gojsonschema_test
+
+import (
+	"errors"
+	"fmt"
+	"testing"
+
+	gjs "github.com/juju/gojsonschema"
+	"github.com/stretchr/testify/assert"
+)
+
+// Get a *Schema given a properties map.
+func schemaFromProperties(t *testing.T, properties map[string]interface{}) *gjs.Schema {
+	schemaMap := map[string]interface{}{
+		"type":       "object",
+		"properties": properties,
+	}
+	loader := gjs.NewGoLoader(schemaMap)
+	schema, err := gjs.NewSchema(loader)
+	assert.Nil(t, err)
+	return schema
+}
+
+func testGetDocProperties(schema *gjs.Schema) (doc map[string]interface{}, testError error) {
+	// Capture expected panics
+	defer func() {
+		if r := recover(); r != nil {
+			var msg string
+			// This should never panic
+			switch typed := r.(type) {
+			case error:
+				msg = fmt.Sprintf("panic: %s", typed.Error())
+			default:
+				msg = fmt.Sprintf("unexpected panic: %#v", typed)
+			}
+			testError = errors.New(msg)
+		}
+	}()
+	schemaDoc := schema.GetDocProperties()
+	return schemaDoc, nil
+}
+
+func TestGetDocProperties(t *testing.T) {
+	for i, test := range []struct {
+		should         string
+		usingSchema    interface{}
+		expectedErr    string
+		expectedResult interface{}
+	}{{
+		should:      "panic for nil pool",
+		expectedErr: "runtime error: invalid memory address or nil pointer dereference",
+	}, {
+		should:      "panic for non-map document",
+		usingSchema: "not-a-map",
+		expectedErr: "interface conversion: interface is string, not map[string]interface {}",
+	}, {
+		should:      "panic for non schema document",
+		usingSchema: map[string]interface{}{"foo": "bar"},
+		expectedErr: "interface conversion: interface is nil, not map[string]interface {}",
+	}, {
+		should: "work for an OK document",
+		usingSchema: map[string]interface{}{
+			"properties": map[string]interface{}{
+				"foo": "bar"}},
+		expectedResult: map[string]interface{}{"foo": "bar"},
+	}} {
+		fmt.Printf("TestGetDocProperties test (%d) | should %s\n", i, test.should)
+		schema := gjs.MakeTestingSchema(test.usingSchema)
+		doc, err := testGetDocProperties(schema)
+		if test.expectedErr != "" {
+			assert.EqualError(t, err, "panic: "+test.expectedErr)
+			continue
+		}
+		assert.NoError(t, err)
+		assert.Equal(t, test.expectedResult, doc)
+	}
+
+	println()
+}
+
+func TestIterateAndInsert(t *testing.T) {
+	for i, test := range []struct {
```

Comment:
Good tests, but hard to follow. Separating tests and loop would help. Also a helper creating those `map[string]interface{}`, e.g. a `func M(in ...interface{}) map[string]interface{}` where every first value is type asserted as `string` (the key) and every second as its value.

```
myMap := M(
    "Foo", 42,
    "Bar", M("Baz", 1, "Woz", false),
)
```